Framework: Upgrade to Babel 6

aduth opened the pull request:
This pull request seeks to upgrade Calypso from Babel 5.8.12 to Babel 6.9.x. [Babel 6](http://babeljs.io/blog/2015/10/29/6.0.0/) is a significant departure in its approach, switching to a modular approach. As such, it requires opting in to most syntax conversions.

A significant number of the changed files are a consequence of the splitting of `babel/register` to a separate [`babel-register` module](https://babeljs.io/docs/usage/require/). Since most of our tests use the require hook, each of their Makefiles needed to be updated to reflect the updated compiler module.

__Testing instructions:__

Run `make test` from the root directory and verify that all tests pass.

Ensure that regressions should be found in navigating the application in the browser.

aduth commented:
Flipping this back to "In Progress", as there appears to be some issues related to `babel-runtime` which I'm not able to reproduce locally, but are causing test failures in CircleCI. Adding `babel-runtime` as a dependency results in a different error which looks very similar to the one reported here:

https://phabricator.babeljs.io/T6644

Will plan to take a closer look soon, or wait until a proper resolution is put in place in Babel if it indeed turns out to be an issue there.

aduth commented:
The differences between what I was seeing locally compared to CircleCI appear to be due to the fact that I was running a more recent version of `node` / `npm`. In which case, upgrading our Node dependency in #1204 could be a solution, though I'll see if there's another interim option.

aduth commented:
I've rebased and the tests now _run_, probably due to Node 4 in #1204 (though we're still on npm 2.x).

However, the tests still fail consistently, due to a few slow test suites. I'm unsure yet whether this is an issue with the tests themselves (FeedStore and FeedPostList), or if Babel 6 is actually slower than Babel 5, causing the tests to fall over the timeout.

/cc @blowery

blowery commented:
hmmmm

```
FeedStore
    ✓ should have a dispatch token
    ✓ should not contain an unknown feed ID
    ✓ should create a pending record when sent REFRESH
    ✓ should pass through the pending state and prevent double fetches
    ✓ should accept a good response
    ✓ should accept an error


  6 passing (664ms)



  FeedPostList
    ✓ should require an id, a fetcher, a keyMaker
    A valid instance
      ✓ should receive a page
      ✓ should receive updates
      ✓ should treat each set of updates as definitive
    Selected index
      ✓ should initially have nothing selected
      ✓ should select the next item
      ✓ should select the next valid post
      ✓ should be able to select a gap
      ✓ should select the prev item
      ✓ should select the prev valid post
    Filter followed x-posts
      ✓ rolls up x-posts and matching x-comments
      ✓ when following origin site, filters followed x-posts, but leaves comment notices
      ✓ updates sites x-posted to
      ✓ filters xposts with no metadata
      ✓ filters xposts with missing xpost metadata


  15 passing (938ms)
```

blowery commented:
though i'm on npm 3.something...

aduth commented:
I've seen at least one report claiming that Babel can be significantly slower when used with npm 2 than with npm 3. Not obvious to me why this would be the case (related to flattened directly structure?):

https://news.ycombinator.com/item?id=10475713

I'll plan to do some more testing early next week.

blowery commented:
https://github.com/este/este/commit/b04f95114ca13e994b4c2d2207e00b907275dcc1 might help

aduth commented:
>este/este@b04f951 might help

Definitely worth testing to see if it helps. I'll commit shortly.

blowery commented:
rebased this on current master, fixed the conflicts, added the npm3 to see.

if npm3 doesn't fix it, we can back it out pretty easily.

blowery commented:
Just a quick data point: building `make build-wpcalypso` takes 40s on both babel 5 (40.2s) and babel 6 (40.4s).

dmsnell commented on client/lib/site/Makefile:
```diff
@@ -2,7 +2,7 @@ NODE_BIN := $(shell npm bin)
 MOCHA ?= $(NODE_BIN)/mocha
 BASE_DIR := $(NODE_BIN)/../..
 NODE_PATH := test:$(BASE_DIR)/client:$(BASE_DIR)/shared
-COMPILERS ?= js:babel/register-without-polyfill
```
I test this just to make sure it didn't break without the polyfill version and it was fine - 24/24 passing

dmsnell commented on client/lib/upgrades/actions/index.js:
```diff
@@ -1,5 +1,13 @@
+	...domainManagement,
+	...domainSearch,
+	...purchases
+};
```
Is this supposed to be included in this PR?

aduth replied:
>Is this supposed to be included in this PR?

I'll look at this a bit more. It curently fails to compile with the following error with the original code:

```
ERROR in ./client/lib/upgrades/actions/index.js
Module parse failed: /Users/andrew/Documents/Code/wp-calypso/node_modules/react-hot-loader/index.js!/Users/andrew/Documents/Code/wp-calypso/node_modules/babel-loader/index.js?cacheDirectory!/Users/andrew/Documents/Code/wp-calypso/client/lib/upgrades/actions/index.js Line 5: Unexpected token
You may need an appropriate loader to handle this file type.
| 'use strict';
| 
| import _Object$defineProperty from 'babel-runtime/core-js/object/define-property';
| Object.defineProperty(exports, "__esModule", {
|   value: true
 @ ./client/my-sites/current-site/index.jsx 26:22-53
```

However, it's a bit puzzling, since the additional export supports are from a [stage 1 proposal](https://github.com/leebyron/ecmascript-more-export-from) ([source](https://github.com/tc39/ecma262)), so I'd expect them to be supported [per our configuration](https://github.com/Automattic/wp-calypso/blob/d4669869a502b040db6b83694f65019e0e7cac7f/.babelrc#L4).

aduth replied:
Looks like this is a bug with Babel: http://phabricator.babeljs.io/T6888

dmsnell commented on client/me/notification-settings/settings-form/stream.jsx:
```diff
@@ -65,7 +65,7 @@ export default React.createClass( {
 
 					return ( <StreamHeader stream={ this.props.stream } /> );
-				}() }
+				} )() }
```
Nice catch on these - were they breaking tests?

aduth replied:
>Nice catch on these - were they breaking tests?

If I recall correctly, the code was failing to compile on the newer version of Babel.

dmsnell commented:
Looks good @aduth - what obstacles remain before merging this?

aduth commented:
>what obstacles remain before merging this?

The only obstacle was holiday time :smile: I'll respond to your comments shortly.

aduth commented:
Current status:

- Would like to wait for an outstanding Babel issue to be resolved. See [T6888](http://phabricator.babeljs.io/T6888) ([T2877](https://phabricator.babeljs.io/T2877))
- Need to rebase to (a) resolve conflicts and (b) replace any new occurrences of `babel/register` with `babel-register`

blowery commented:
> Also, we should consider the implications of requiring NPM 3 in CircleCI. Should we force/encourage this for local development and/or production environments to ensure consistent installation behavior?

Does using npm3 affect bundle size too? Might be beneficial to force it.

aduth commented:
>Does using npm3 affect bundle size too?

I wouldn't think that it should, but then again, I wouldn't have expected it to have any impact on build speed. My worry is more the potential for false positives in CircleCI due to running a different version of npm. If the impact on build speed is significant enough, we would certainly want to encourage it for the developer's benefit.

aduth commented:
@jkudish @coderkevin : This pull request is now once-again up-to-date, and is much more digestable since the elimination of individual Makefiles with our single test runner migration (#3942, cc @gziolo) and having upgraded to Node 5.x.

That said, there are still a few hurdles to moving forward with these changes:

**[`rewire`](https://github.com/jhnns/rewire) does not work with Babel 6 (jhnns/rewire#62)**, and it appears that the author does not have an intention of fixing it. We use `rewire` in about a dozen tests, usually to modify or retrieve the value of a module-level "private" variable.

- Are there alternatives? I've found [`babel-plugin-rewire`](https://github.com/speedskater/babel-plugin-rewire), but it requires a different test set-up and I ran into several errors in trying to get it working.
- Should we drop our dependency on `rewire`? As mentioned, there are about a dozen tests which would need to be migrated. Some are fairly straight forward in that we could simply export the module variable (efc6fe1), but a few others will require some more thought. In my opinion, `rewire` has always felt a bit of a hack and I feel partly responsible for its introduction 😬 

**Performance** The gap seems to have narrowed, which is expected, since some performance issues were reported around Babel 6 and npm 2.x dependency structure ([T6756](https://phabricator.babeljs.io/T6756)). That said, my (albeit unscientific) testing shows that it's still about the same or slower than Babel 5.

- See [timings for a handful of runs under both `master` and this branch](https://gist.github.com/aduth/03115e9b8eed6668dfa6fc333271a5d1). Testing times were not reported because I suspect failures in this branch are causing some tests to exit early.
- Why are we finding build times to be slow? Especially when both Babel [6.0.0](https://babeljs.io/blog/2015/10/29/6.0.0#performance-improvements) and [6.5.0](https://github.com/babel/babel/blob/master/CHANGELOG.md#650-2016-02-07) claimed _huge_ performance benefits
 - Is our configuration sub-optimal? Are these claimed performance benefits applicable only to applications smaller than ours?

**Bugs**. Not quite as critical, but it's disappointing that bugs like [T2877](https://phabricator.babeljs.io/T2877) have remained unresolved (since November in this instance).

My overall feeling toward upgrading is that we've not encountered a specific need in Calypso for Babel 6 features, but that it'll become increasingly difficult to upgrade if we let it linger too long. In my opinion, slowed performance should be considered a blocker for upgrading.

dmsnell commented:
@aduth is there something we can add into our linting tests to identify and report code that would prevent an upgrade?

jkudish commented:
For what it's worth, I don't notice any performance difference between the master and this branch, but as you said it's not scientific

gziolo commented:
> Should we drop our dependency on rewire? As mentioned, there are about a dozen tests which would need to be migrated. Some are fairly straight forward in that we could simply export the module variable (efc6fe1), but a few others will require some more thought. In my opinion, rewire has always felt a bit of a hack and I feel partly responsible for its introduction 

It seems like using `rewire` is anti-pattern by design in the terms of unit testing, because code under test shouldn't have side effects :)
I skimmed our code and checked places where we use this library and it looks like in 90% of cases we override private variables of stores or actions(!?). It looks like we can provide simple solution and expose `clean` method for each store/actions file and use it in tests instead. In the long run we are going to move this code to `Redux` anyway. List of files where we use `__set__` rewire method:
![screen shot 2016-04-26 at 09 40 13](https://cloud.githubusercontent.com/assets/699132/14810504/95ffe2d8-0b93-11e6-9850-0f85a8df62a6.png)

aduth commented:
>is there something we can add into our linting tests to identify and report code that would prevent an upgrade?

@dmsnell Probably not worth the effort. The issues are at least fairly easy to identify from what I've experienced, in that the application will either not build or the tests will fail.

>in 90% of cases we override private variables of stores or actions

For my own stores where I've used this approach, this was early in the adoption of "unidirectional data flow" and there was a worry that if the internal storage reference was exposed, developers may try to use and/or manipulate its value. Hence the need to use rewire to reset the value between each test.

>List of files where we use `__set__` rewire method: List of files where use `__get__`:

There's also `__with__`, but probably used in only a couple places if my memory serves correctly.

>`Concluding`: I would drop `rewire` dependency and refactor code.

I'd agree with this 😄

gziolo commented:
I opened PR #5040, where I started removing `rewire` dependency :) In most cases it should be very straightforward. So far I only had some issues with `Dispatcher` dependency which behaved strangly. I was able to resolve it by using `useMockery` helper. I did it to enforce usage of fresh instance of `Dispatcher` in the problematic test file.

gziolo commented:
Drop `rewire` dependency checklist:
* [x] `client/lib/media`
* [ ] `client/lib/post-formats`
* [ ] `client/lib/posts`
* [ ] `client/lib/shortcode`
* [ ] `client/lib/terms`
* [ ] `client/lib/wp/localization`
* [ ] `client/signup`

aduth commented:
Pushed a rebase to resolve merge conflicts. Expected that tests will continue to fail however, pending remaining `rewire` dependencies to be removed (#5810, #5809 #5811).

Remaining on Babel 5 is becoming a bit more painful over time, as plugins we'd like to adopt or create are only available on Babel 6. For example:

- [`babel-plugin-transform-react-remove-prop-types`](https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types) to remove propTypes from production builds
- Pending changes to state selectors to import from single file `import { getSelectedSiteId } from 'state/selectors';` could be optimized with a Babel plugin to avoid unnecessary importing without need for tree shaking from Webpack 2 ([gist](https://gist.github.com/aduth/94d4014b3b96aa15070cb0cab48dfb32))

gziolo commented:
I merged and deployed all `rewire` related PRs. Great work on that one @aduth :)
I removed no longer used `rewire` dependency and regenerated shrinkwrap config again. I had also to resolve two more issues related to Babel upgrade and unit tests. It looks like unit tests are all green now 🎉  🎈
It looks like the tests are a bit slower with the latests Babel version, but I wouldn't consider it as a blocker.
I also checked couple of pages and they work as before.

blowery commented on npm-shrinkwrap.json:
```diff
@@ -134,30 +131,34 @@
     },
     "babel": {
-      "version": "5.8.12",
+      "version": "5.8.38",
```
Who is bringing this in still...

gwwar replied:
It's a dependency of wpcom.

blowery replied:
huh.

blowery replied:
ugh https://github.com/Automattic/wpcom.js/blob/master/dist/index.js#L1

blowery commented on npm-shrinkwrap.json:
```diff
@@ -1363,7 +1573,7 @@
     },
     "graceful-fs": {
-      "version": "4.1.4"
+      "version": "1.2.3"
```
Hmmm. That's ancient. And now the default?

gwwar replied:
Hmm, not sure why npm picked that version over 4.1.4. Maybe it tries to choose the lowest version as root? https://docs.npmjs.com/how-npm-works/npm3-dupe

```
>$ npm ls graceful-fs
wp-calypso@0.17.0 /Users/kerryliu/checkouts/wp-calypso
├─┬ eslint@1.10.3
│ └─┬ file-entry-cache@1.2.4
│   └─┬ flat-cache@1.0.10
│     ├── graceful-fs@4.1.4 
│     └─┬ read-json-sync@1.1.1
│       └── graceful-fs@4.1.4 
├─┬ mocha@2.3.4
│ └─┬ glob@3.2.3
│   └── graceful-fs@2.0.3 
├─┬ node-sass@3.4.2
│ ├─┬ gaze@0.5.2
│ │ └─┬ globule@0.1.0
│ │   └─┬ glob@3.1.21
│ │     └── graceful-fs@1.2.3 
│ ├─┬ meow@3.7.0
│ │ └─┬ read-pkg-up@1.0.1
│ │   └─┬ read-pkg@1.1.0
│ │     ├─┬ load-json-file@1.1.0
│ │     │ └── graceful-fs@4.1.4 
│ │     └─┬ path-type@1.1.0
│ │       └── graceful-fs@4.1.4 
│ └─┬ node-gyp@3.3.1
│   ├─┬ fstream@1.0.9
│   │ └── graceful-fs@4.1.4 
│   └── graceful-fs@4.1.4 
├─┬ postcss-cli@2.5.1
│ └─┬ chokidar@1.5.1
│   ├─┬ fsevents@1.0.12
│   │ └─┬ node-pre-gyp@0.6.25
│   │   └─┬ tar@2.2.1
│   │     └─┬ fstream@1.0.8
│   │       └── graceful-fs@4.1.3 
│   └─┬ readdirp@2.0.0
│     └── graceful-fs@4.1.4 
├─┬ webpack@1.12.15
│ ├─┬ enhanced-resolve@0.9.1
│ │ └── graceful-fs@4.1.4 
│ └─┬ watchpack@0.2.9
│   └── graceful-fs@4.1.4 
└─┬ wpcom@4.9.13
  └─┬ babel@5.8.38
    └─┬ babel-core@5.8.38
      └─┬ regenerator@0.8.40
        └─┬ commoner@0.10.4
          └── graceful-fs@4.1.4 
```

blowery replied:
ah, yeah, node-sass is the main culprit. looks like that will get bumped when we move to node 6 in #5595 so yay

gwwar commented:
👍 I tested this against a local docker 🐋 image. No visible changes to Calypso.

gwwar commented:
@aduth were there any other blockers on this one? The branch needs a rebase, but I think it's really close.

aduth commented:
@gwwar Yeah, it's essentially mergeable. Performance hit is really unfortunate but I don't think we'll be able to stick to Babel 5 long-term, so the hit is inevitable.

The only other thing I wanted to check to see was whether there is any impact consideration for other projects that depend on Calypso, such as the desktop app.

aduth commented:
Rebased to resolve merge conflicts. Also made the following changes:

- Changed `stage-1` preset to `stage-2`
 - Separately includes [`transform-class-properties`](https://babeljs.io/docs/plugins/transform-class-properties/) ([ES Class Fields & Static Properties](https://github.com/jeffmo/es-class-fields-and-static-properties)) and [`transform-export-extensions`](https://babeljs.io/docs/plugins/transform-export-extensions/) ([export ns from](https://github.com/leebyron/ecmascript-export-ns-from), [export default from](https://github.com/leebyron/ecmascript-export-default-from)) plugins
 - Leaves out decorators and class constructor calls from stage 1 ([reference](https://github.com/babel/babel/blob/master/packages/babel-preset-stage-1/index.js)), which are not currently used
- Dropped `babel-preset-react` in favor of granular plugins
 - Using [`syntax-jsx`](https://babeljs.io/docs/plugins/syntax-jsx/), [`transform-react-jsx`](https://babeljs.io/docs/plugins/transform-react-jsx/), and [`transform-react-display-name`](https://babeljs.io/docs/plugins/transform-react-display-name/) plugins
 - __*Not*__ using [`syntax-flow`](http://babeljs.io/docs/plugins/syntax-flow/) nor [`transform-flow-strip-types`](http://babeljs.io/docs/plugins/transform-flow-strip-types/) plugins

Hope with these changes is (a) limit the number of dependencies, (b) potentially gain some performance boost (though in my anecdotal experience there is no noticeable difference) and (c) avert temptation to use stage 1 features or Flow types.

gwwar commented:
:shipit:
